improver_tests/categorical/decision_tree/test_ApplyDecisionTree.py

```diff
@@ -81,20 +81,90 @@ def hail_cube() -> Cube:
     return cube
 
 
+@pytest.fixture()
+def cloud_top_temp() -> Cube:
+    """
+    Set up a cloud top temperature, deterministic cube.
+    """
+    data = np.full((7, 3), dtype=np.float32, fill_value=250)
+    cube = set_up_variable_cube(
+        data,
+        name="cloud_top_temperature",
+        units="K",
+        time=dt(2017, 10, 10, 12, 0),
+        frt=dt(2017, 10, 10, 11, 0),
+    )
+    return cube
+
+
+@pytest.fixture()
+def cloud_base_temp() -> Cube:
+    """
+    Set up a cloud base temperature, deterministic cube.
+    """
+    data = np.full((7, 3), dtype=np.float32, fill_value=270)
+    cube = set_up_variable_cube(
+        data,
+        name="cloud_base_temperature",
+        units="K",
+        time=dt(2017, 10, 10, 12, 0),
+        frt=dt(2017, 10, 10, 11, 0),
+    )
+    return cube
+
+
+@pytest.mark.parametrize(
+    "masked_data", ("no_mask", "mask_precip", "mask_cloud_top", "mask_cloud_base")
+)
 @pytest.mark.parametrize(
     "precip_fill,hail_fill,expected", ((1, 1, 2), (1, 0, 1), (0, 0, 0), (0, 1, 0))
 )
 def test_non_probablistic_tree(
-    precip_cube, hail_cube, precip_fill, hail_fill, expected
+    precip_cube,
+    hail_cube,
+    cloud_top_temp,
+    cloud_base_temp,
+    precip_fill,
+    hail_fill,
+    expected,
+    masked_data,
 ):
     """Test that ApplyDecisionTree correctly manages a decision tree with deterministic
-    inputs"""
+    inputs. Also check that if data is masked the output will be masked unless the
+    node includes an if_masked setting."""
+
+    expected_array = np.full_like(precip_cube.data, expected)
+
+    if masked_data == "mask_precip":
+        # Test that if there is no if_masked setting in the decision tree, the output
+        # will be masked
+        precip_mask = np.full_like(precip_cube.data, False)
+        precip_mask[0, 0] = True
+        precip_cube.data = np.ma.masked_array(precip_cube.data, mask=precip_mask)
+
+        expected_mask = np.full_like(precip_cube.data, False)
+        expected_mask[0, 0] = True
+        expected = np.ma.masked_array(expected_array, mask=expected_mask)
```

Comment:
This is suspicious. The test at the end looks at `expected_array` and yet this test still passes. This problem is hidden by another problem in the test at the end - see later comment.

Reply:
I've updated the decision tree slightly so the if masked will go to false rather than true which would stop this test passing.